### DISCUSSION TITLE
How to use WGLMakie in Jupyter Notebook through ssh?

### DISCUSSION THREAD

**Discussion by aramirezreyes** [JuliaPlots/Makie.jl#4182] [Q&A]
This is a follow up to a discussion from slack (actually modifying the use case), but I think it deserves to be here so that we can work on it on an organized manner. 

Case: I connect to a remote server via ssh. I launch a Jupyter notebook instance in this server, this goes back to my computer via the ssh channel. 
```
Local machine:
  ssh -L 8889:localhost:8889 user@server -Y -X
At server:
  jupyter-notebook --no-browser --port=8889
Local machine's web browser:
  localhost:8889
```

Then, inside the jupyter notebook I try to use WGLMakie:

```
julia> using AbstractPlotting, WGLMakie; AbstractPlotting.inline!(true)
julia> plot(rand(10))
```

Result: 
I get no plot, instead I get an error 

> Firefox can't establish a connection to the server at 127.0.0.1:8081
> The site could be temporarily unavailable or too busy. Try again in a few moments.
> If you are unable to load any pages, check your computer's network connection.
> If your computer or network is protected by a firewall or proxy, make sure that Firefox is permitted to access the Web.

Desired result: I see a nice looking plot in my browser.

How could I make this work?

**Comment by SimonDanisch** [comment]
With a bit of luck, changing the port to 8081 could make things work

**Comment by aramirezreyes** [comment]
Bad luck! Now the content was blocked by security policy. I am not sure of closing this or leaving it open so that someone else can find it when looking for a similar thing.

**Comment by SimonDanisch** [comment]
Ah, but I guess that means it works almost and should be fixable somehow!
Try setting `ENV["WEBIO_HTTP_PORT"] = 8889` before using WGMakie and then use that port - this may help if only that specific port is unblocked...
Otherwise, it'd be nice to see the full error, maybe with a screenshot!

**Comment by aramirezreyes** [comment]
Great! I feel like it is almost there:

I am seeing an infitine loop of this:
![image](https://user-images.githubusercontent.com/6992761/73777393-302cac80-473e-11ea-9c39-1220cde65a7d.png)
in the Jupyter console
after doing `plot(rand(10))`,
but the cell never finishes evaluating.

**Comment by cossio** [comment]
Did you figure this out? I have same issue. I have to use WGLMakie since GLMakie doesn't work on the server due to some OpenGL issue.

**Comment by SimonDanisch** [comment]
Jupyter or Pluto?
With Pluto, this should work: https://makie.juliaplots.org/stable/documentation/remote/#wglmakie
IJulia is broken right now - i'm 90% done with a fix...

**Comment by cossio** [comment]
I use Jupyter. But I'll try Pluto.
Hope the fix gets there soon! Thanks

**Comment by cossio** [comment]
Any updates? This is what I am getting right now:

![image](https://user-images.githubusercontent.com/10637482/142863862-c501f032-d0b4-4621-aad2-0853aea1fdb4.png)

No plot :disappointed: 

CairoMakie works fine, though.

Pluto is not always an option, and CairoMakie is slow for large datasets. It would be nice to get this fixed for Jupyter.

**Comment by SimonDanisch** [comment]
I'm working on it: https://github.com/SimonDanisch/JSServe.jl/pull/111
Maybe I can get it merged this week.

**Comment by cossio** [comment]
Nice, thanks. @SimonDanisch if that PR gets merged, it means that it will just work, without any extra config?

**Comment by SimonDanisch** [comment]
I'm aiming for that and it should be possible, so yes! Lets hope nothing too unexpected jinxes it :D

**Comment by cossio** [comment]
Good! Looking forward to it.

**Comment by frankier** [comment]
This should help https://github.com/MakieOrg/Makie.jl/issues/2405

**Comment by ffreyer** [comment]
Should we move this to Q&A?
